File: backend/routers/trucks.py
import sqlite3
from typing import Optional
from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel
from database import get_db
from auth import require_owner

router = APIRouter()


class TruckCreate(BaseModel):
    make: str
    model: str
    year: int
    license_plate: str
    capacity_tons: float
    truck_type: str
    price_per_day: float
    description: Optional[str] = None
    location: Optional[str] = None


class TruckUpdate(BaseModel):
    make: Optional[str] = None
    model: Optional[str] = None
    year: Optional[int] = None
    license_plate: Optional[str] = None
    capacity_tons: Optional[float] = None
    truck_type: Optional[str] = None
    price_per_day: Optional[float] = None
    description: Optional[str] = None
    location: Optional[str] = None
    is_available: Optional[bool] = None


@router.get("")
def list_trucks(
    type: Optional[str] = None,
    min_price: Optional[float] = None,
    max_price: Optional[float] = None,
    location: Optional[str] = None,
    available: Optional[str] = None,
    db: sqlite3.Connection = Depends(get_db),
):
    query = """
        SELECT t.*, u.name as owner_name, u.phone as owner_phone,
            COALESCE(AVG(r.rating), 0) as avg_rating,
            COUNT(r.id) as review_count
        FROM trucks t
        JOIN users u ON t.owner_id = u.id
        LEFT JOIN reviews r ON r.truck_id = t.id
        WHERE 1=1
    """
    params = []
    if type:
        query += " AND t.truck_type = ?"
        params.append(type)
    if min_price is not None:
        query += " AND t.price_per_day >= ?"
        params.append(min_price)
    if max_price is not None:
        query += " AND t.price_per_day <= ?"
        params.append(max_price)
    if location:
        query += " AND t.location LIKE ?"
        params.append(f"%{location}%")
    if available is not None:
        query += " AND t.is_available = ?"
        params.append(1 if available == "true" else 0)

    query += " GROUP BY t.id ORDER BY t.created_at DESC"
    rows = db.execute(query, params).fetchall()
    return [dict(r) for r in rows]


@router.get("/owner/my-trucks")
def my_trucks(user: dict = Depends(require_owner), db: sqlite3.Connection = Depends(get_db)):
    rows = db.execute(
        """
        SELECT t.*,
            COALESCE(AVG(r.rating), 0) as avg_rating,
            COUNT(DISTINCT r.id) as review_count,
            COUNT(DISTINCT b.id) as booking_count
        FROM trucks t
        LEFT JOIN reviews r ON r.truck_id = t.id
        LEFT JOIN bookings b ON b.truck_id = t.id
        WHERE t.owner_id = ?
        GROUP BY t.id
        ORDER BY t.created_at DESC
        """,
        (user["id"],),
    ).fetchall()
    return [dict(r) for r in rows]


@router.get("/{truck_id}")
def get_truck(truck_id: int, db: sqlite3.Connection = Depends(get_db)):
    row = db.execute(
        """
        SELECT t.*, u.name as owner_name, u.phone as owner_phone, u.email as owner_email,
            COALESCE(AVG(r.rating), 0) as avg_rating,
            COUNT(r.id) as review_count
        FROM trucks t
        JOIN users u ON t.owner_id = u.id
        LEFT JOIN reviews r ON r.truck_id = t.id
        WHERE t.id = ?
        GROUP BY t.id
        """,
        (truck_id,),
    ).fetchone()
    if not row:
        raise HTTPException(404, "Truck not found")
    return dict(row)


@router.post("", status_code=201)
def create_truck(body: TruckCreate, user: dict = Depends(require_owner), db: sqlite3.Connection = Depends(get_db)):
    existing = db.execute("SELECT id FROM trucks WHERE license_plate = ?", (body.license_plate,)).fetchone()
    if existing:
        raise HTTPException(409, "License plate already registered")

    cur = db.execute(
        """
        INSERT INTO trucks (owner_id, make, model, year, license_plate, capacity_tons,
                            truck_type, price_per_day, description, location)
        VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
        """,
        (user["id"], body.make, body.model, body.year, body.license_plate,
         body.capacity_tons, body.truck_type, body.price_per_day, body.description, body.location),
    )
    db.commit()
    row = db.execute("SELECT * FROM trucks WHERE id = ?", (cur.lastrowid,)).fetchone()
    return dict(row)


@router.put("/{truck_id}")
def update_truck(
    truck_id: int,
    body: TruckUpdate,
    user: dict = Depends(require_owner),
    db: sqlite3.Connection = Depends(get_db),
):
    row = db.execute("SELECT * FROM trucks WHERE id = ?", (truck_id,)).fetchone()
    if not row:
        raise HTTPException(404, "Truck not found")
    if row["owner_id"] != user["id"]:
        raise HTTPException(403, "Not your truck")

    is_available_val = None
    if body.is_available is not None:
        is_available_val = 1 if body.is_available else 0

    db.execute(
        """
        UPDATE trucks SET
            make = COALESCE(?, make),
            model = COALESCE(?, model),
            year = COALESCE(?, year),
            license_plate = COALESCE(?, license_plate),
            capacity_tons = COALESCE(?, capacity_tons),
            truck_type = COALESCE(?, truck_type),
            price_per_day = COALESCE(?, price_per_day),
            description = COALESCE(?, description),
            location = COALESCE(?, location),
            is_available = COALESCE(?, is_available)
        WHERE id = ?
        """,
        (body.make, body.model, body.year, body.license_plate, body.capacity_tons,
         body.truck_type, body.price_per_day, body.description, body.location,
         is_available_val, truck_id),
    )
    db.commit()
    updated = db.execute("SELECT * FROM trucks WHERE id = ?", (truck_id,)).fetchone()
    return dict(updated)


@router.delete("/{truck_id}")
def delete_truck(truck_id: int, user: dict = Depends(require_owner), db: sqlite3.Connection = Depends(get_db)):
    row = db.execute("SELECT * FROM trucks WHERE id = ?", (truck_id,)).fetchone()
    if not row:
        raise HTTPException(404, "Truck not found")
    if row["owner_id"] != user["id"]:
        raise HTTPException(403, "Not your truck")

    db.execute("DELETE FROM trucks WHERE id = ?", (truck_id,))
    db.commit()
    return {"message": "Truck deleted"}

File: backend/routes/trucks.js
const express = require('express');
const db = require('../db');
const { authenticate, requireOwner } = require('../middleware/auth');

const router = express.Router();

router.get('/', (req, res) => {
  const { type, min_price, max_price, location, available } = req.query;
  let query = `
    SELECT t.*, u.name as owner_name, u.phone as owner_phone,
      COALESCE(AVG(r.rating), 0) as avg_rating,
      COUNT(r.id) as review_count
    FROM trucks t
    JOIN users u ON t.owner_id = u.id
    LEFT JOIN reviews r ON r.truck_id = t.id
    WHERE 1=1
  `;
  const params = [];

  if (type) { query += ' AND t.truck_type = ?'; params.push(type); }
  if (min_price) { query += ' AND t.price_per_day >= ?'; params.push(Number(min_price)); }
  if (max_price) { query += ' AND t.price_per_day <= ?'; params.push(Number(max_price)); }
  if (location) { query += ' AND t.location LIKE ?'; params.push(`%${location}%`); }
  if (available !== undefined) { query += ' AND t.is_available = ?'; params.push(available === 'true' ? 1 : 0); }

  query += ' GROUP BY t.id ORDER BY t.created_at DESC';

  const trucks = db.prepare(query).all(...params);
  res.json(trucks);
});

router.get('/:id', (req, res) => {
  const truck = db.prepare(`
    SELECT t.*, u.name as owner_name, u.phone as owner_phone, u.email as owner_email,
      COALESCE(AVG(r.rating), 0) as avg_rating,
      COUNT(r.id) as review_count
    FROM trucks t
    JOIN users u ON t.owner_id = u.id
    LEFT JOIN reviews r ON r.truck_id = t.id
    WHERE t.id = ?
    GROUP BY t.id
  `).get(req.params.id);
  if (!truck) return res.status(404).json({ error: 'Truck not found' });
  res.json(truck);
});

router.post('/', authenticate, requireOwner, (req, res) => {
  const { make, model, year, license_plate, capacity_tons, truck_type, price_per_day, description, location } = req.body;
  if (!make || !model || !year || !license_plate || !capacity_tons || !truck_type || !price_per_day) {
    return res.status(400).json({ error: 'make, model, year, license_plate, capacity_tons, truck_type, and price_per_day are required' });
  }

  const existing = db.prepare('SELECT id FROM trucks WHERE license_plate = ?').get(license_plate);
  if (existing) return res.status(409).json({ error: 'License plate already registered' });

  const result = db.prepare(`
    INSERT INTO trucks (owner_id, make, model, year, license_plate, capacity_tons, truck_type, price_per_day, description, location)
    VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
  `).run(req.user.id, make, model, Number(year), license_plate, Number(capacity_tons), truck_type, Number(price_per_day), description || null, location || null);

  const truck = db.prepare('SELECT * FROM trucks WHERE id = ?').get(result.lastInsertRowid);
  res.status(201).json(truck);
});

router.put('/:id', authenticate, requireOwner, (req, res) => {
  const truck = db.prepare('SELECT * FROM trucks WHERE id = ?').get(req.params.id);
  if (!truck) return res.status(404).json({ error: 'Truck not found' });
  if (truck.owner_id !== req.user.id) return res.status(403).json({ error: 'Not your truck' });

  const { make, model, year, license_plate, capacity_tons, truck_type, price_per_day, description, location, is_available } = req.body;

  db.prepare(`
    UPDATE trucks SET
      make = COALESCE(?, make),
      model = COALESCE(?, model),
      year = COALESCE(?, year),
      license_plate = COALESCE(?, license_plate),
      capacity_tons = COALESCE(?, capacity_tons),
      truck_type = COALESCE(?, truck_type),
      price_per_day = COALESCE(?, price_per_day),
      description = COALESCE(?, description),
      location = COALESCE(?, location),
      is_available = COALESCE(?, is_available)
    WHERE id = ?
  `).run(make, model, year ? Number(year) : null, license_plate, capacity_tons ? Number(capacity_tons) : null,
    truck_type, price_per_day ? Number(price_per_day) : null, description, location,
    is_available !== undefined ? (is_available ? 1 : 0) : null, req.params.id);

  const updated = db.prepare('SELECT * FROM trucks WHERE id = ?').get(req.params.id);
  res.json(updated);
});

router.delete('/:id', authenticate, requireOwner, (req, res) => {
  const truck = db.prepare('SELECT * FROM trucks WHERE id = ?').get(req.params.id);
  if (!truck) return res.status(404).json({ error: 'Truck not found' });
  if (truck.owner_id !== req.user.id) return res.status(403).json({ error: 'Not your truck' });

  db.prepare('DELETE FROM trucks WHERE id = ?').run(req.params.id);
  res.json({ message: 'Truck deleted' });
});

router.get('/owner/my-trucks', authenticate, requireOwner, (req, res) => {
  const trucks = db.prepare(`
    SELECT t.*,
      COALESCE(AVG(r.rating), 0) as avg_rating,
      COUNT(DISTINCT r.id) as review_count,
      COUNT(DISTINCT b.id) as booking_count
    FROM trucks t
    LEFT JOIN reviews r ON r.truck_id = t.id
    LEFT JOIN bookings b ON b.truck_id = t.id
    WHERE t.owner_id = ?
    GROUP BY t.id
    ORDER BY t.created_at DESC
  `).all(req.user.id);
  res.json(trucks);
});

module.exports = router;
